# Leetcode Practice

## 122. Best Time to Buy and Sell Stock II

Say you have an array prices for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete as many transactions as you like (i.e., buy one and sell one share of the stock multiple times).

Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).

In [2]:
#function
def maxProfit( prices):
    profit = []
    for i in range(1,len(prices)):
        profit.append(prices[i] - prices[i-1]) #taking the difference

    for i,num in enumerate(profit): #replacing neg differences with zero
        if num <= 0:
            profit[i] = 0
        else:
            profit[i] = num

    return sum(profit)


    #buy when low, sell when high
    #sum all profits
    #Runtime: 76 ms, faster than 34.17% of Python3 online submissions for Best Time to Buy and Sell Stock II.
    #Memory Usage: 15.2 MB, less than 48.17% of Python3 online submissions for Best Time to Buy and Sell Stock II.

In [3]:
#test
test = [7,1,5,3,6,4]
maxProfit(test)

7

Input: [7,1,5,3,6,4]

Output: 7

Explanation: Buy on day 2 (price = 1) and sell on day 3 (price = 5), profit = 5-1 = 4.
             Then buy on day 4 (price = 3) and sell on day 5 (price = 6), profit = 6-3 = 3.

# 189. Rotate Array

Given an array, rotate the array to the right by k steps, where k is non-negative.

Follow up:

Try to come up as many solutions as you can, there are at least 3 different ways to solve this problem.
Could you do it in-place with O(1) extra space?

In [38]:
def rotate(nums, k):
    if k > -1 and k <= len(nums):
        i = abs(len(nums) - k)
        nums.extend(nums[0:i]) #adds the first three numbers to the back
        del nums[0:i] #deletes the same numbers from the front
        return nums 
    elif k > len(nums):
        for i in range(0,k):
            j = nums[-1]
            nums.insert(0,j) #adds 1 to the front of the list
            del nums[-1] #removes the last element from the back of the list
        return nums
    else:
        return "Error: K is too large or non-positive"
    
    #Runtime: 48 ms, faster than 99.86% of Python3 online submissions for Rotate Array.
    #Memory Usage: 15.3 MB, less than 67.33% of Python3 online submissions for Rotate Array.

In [39]:
#test
nums = [1,2,3,4,5,6,7]
k = 3

In [40]:
rotate(nums, k)

[5, 6, 7, 1, 2, 3, 4]

Input: nums = [1,2,3,4,5,6,7], k = 3

Output: [5,6,7,1,2,3,4]

Explanation:

rotate 1 steps to the right: [7,1,2,3,4,5,6]

rotate 2 steps to the right: [6,7,1,2,3,4,5]

rotate 3 steps to the right: [5,6,7,1,2,3,4]

# 350. Intersection of Two Arrays II

Given two arrays, write a function to compute their intersection.

In [15]:
def intersect(nums1, nums2):
    result = [] #creating result array
    for num in nums1:
        if num in nums2:
            result.append(num)
            nums2.remove(num)
    return result
    #Runtime: 132 ms, faster than 5.12% of Python3 online submissions for Intersection of Two Arrays II.
    #Memory Usage: 14.2 MB, less than 7.71% of Python3 online submissions for Intersection of Two Arrays II.

In [16]:
#test
nums1 = [1,2,2,1]
nums2 = [2,2]

In [17]:
intersect(nums1, nums2)

[2, 2]

Input: nums1 = [1,2,2,1], nums2 = [2,2]

Output: [2,2]

# 66. Plus One

Given a non-empty array of digits representing a non-negative integer, increment one to the integer.

The digits are stored such that the most significant digit is at the head of the list, and each element in the array contains a single digit.

You may assume the integer does not contain any leading zero, except the number 0 itself.

In [21]:
def plusOne(digits):
    convert = [str(i) for i in digits]
    squish = "".join(convert)
    num = int(squish)
    new = num + 1
    a_list = [int(x) for x in str(new)]
    return a_list

#Runtime: 56 ms, faster than 10.53% of Python3 online submissions for Plus One.
#Memory Usage: 13.9 MB, less than 39.28% of Python3 online submissions for Plus One.

In [22]:
#test
digits = [1,2,3]
plusOne(digits)

[1, 2, 4]

Input: digits = [1,2,3]
    
Output: [1,2,4]
    
Explanation: The array represents the integer 123.

# 283. Move Zeros


Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.

In [36]:
def moveZeroes(nums):
    for num in nums:
        if num == 0:
            nums.remove(num)
            nums.append(num)
            
#Runtime: 228 ms, faster than 14.21% of Python3 online submissions for Move Zeroes.
#Memory Usage: 14.9 MB, less than 95.23% of Python3 online submissions for Move Zeroes.

In [37]:
#test
test = [0,1,0,3,12]
moveZeroes(test)
print(test)

[1, 3, 12, 0, 0]


Example:

Input: [0,1,0,3,12]

Output: [1,3,12,0,0]

# 344. Reverse String

Write a function that reverses a string. The input string is given as an array of characters char[].

Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.

You may assume all the characters consist of printable ascii characters.

In [53]:
def reverseString(s):
    s.reverse()
    #Runtime: 308 ms, faster than 18.66% of Python3 online submissions for Reverse String.
    #Memory Usage: 18.1 MB, less than 89.58% of Python3 online submissions for Reverse String.

In [54]:
#test
test = ["h","e","l","l","o"]

In [55]:
reverseString(test)

In [56]:
print(test)

['o', 'l', 'l', 'e', 'h']


Example:
    
Input: ["h","e","l","l","o"]
    
Output: ["o","l","l","e","h"]

# 7. Reverse Integer

Given a 32-bit signed integer, reverse digits of an integer.

Note:
Assume we are dealing with an environment which could only store integers within the 32-bit signed integer range: [−231,  231 − 1]. For the purpose of this problem, assume that your function returns 0 when the reversed integer overflows.

In [1]:
def reverse(x):
    if str(x)[0] == '-':
        x = list(str(x))[1:]
        x.reverse()
        a = "".join(x)
        test = int(a) * -1
        if test > -2**31:
            return test
        else:
            return 0
    else:
        x = list(str(x))
        x.reverse()
        a = "".join(x)
        test = int(a)
        if test < ((2**31)-1):
            return test
        else:
            return 0
        
        #Runtime: 56 ms, faster than 10.85% of Python3 online submissions for Reverse Integer.
        #Memory Usage: 13.8 MB, less than 67.51% of Python3 online submissions for Reverse Integer.

In [2]:
x = -123

In [3]:
reverse(x)

-321

Example 2:

Input: -123

Output: -321

# 387. First Unique Character in a String

Given a string, find the first non-repeating character in it and return its index. If it doesn't exist, return -1.

In [7]:
from collections import defaultdict
def firstUniqChar(nums):
    if len(nums) > 0: #if list is not empty
        test = list(nums)
        hash_table = defaultdict(int) #default dict an empty dictionary of integers (ints) that fill all empty keys, so you don't get a                                            #key error
        for i in nums:
            hash_table[i] += 1 #counts how many times each element in the list appears

        for let in hash_table:
            if hash_table[let] == 1: #if value is equal to 1 aka if value has a frequency of 1 in the defaultdict
                #NOTE: this will catch the first unique character bc the hash-table preserves the original list order
                return test.index(let) #print the index of test that corresponds to that value (letter)
        return -1
    else:
        return -1
    #Runtime: 108 ms, faster than 72.69% of Python3 online submissions for First Unique Character in a String.
    #Memory Usage: 14 MB, less than 26.49% of Python3 online submissions for First Unique Character in a String.


In [8]:
test = "loveleetcode"

In [9]:
print(firstUniqChar(test))

2


Examples:

s = "leetcode"
return 0.

s = "loveleetcode"
return 2.